k2/csrc/fsa_utils.cu

```diff
@@ -99,75 +99,241 @@ static void SplitStringToVector(const std::string &in, const char *delim,
   }
 }
 
-// Create an acceptor from a stream.
-static Fsa AcceptorFromStream(std::string first_line, std::istringstream &is,
-                              bool openfst) {
+/* Create an acceptor from a stream, assuming the acceptor is in the K2 format:
+
+   src_state1 dest_state1 label1 score1
+   src_state2 dest_state2 label2 score2
+   ... ...
+   final_state
+
+   The source states will be in non-descending order, and the final state does
+   not bear a cost/score -- we put the cost/score on the arc that connects to
+   the final state and set its label to -1.
+
+   @param [in]  is    The input stream that contains the acceptor.
+
+   @return It returns an Fsa on CPU.
+*/
+static Fsa K2AcceptorFromStream(std::istringstream &is) {
   std::vector<Arc> arcs;
-  std::string line = std::move(first_line);
   std::vector<std::string> splits;
+  std::string line;
 
-  float scale = 1;
-  if (openfst) scale = -1;
+  bool finished = false;  // when the final state is read, set it to true.
+  while (std::getline(is, line)) {
+    SplitStringToVector(line, kDelim,
+                        &splits);  // splits is cleared in the function
+    if (splits.empty()) continue;  // this is an empty line
+
+    K2_CHECK_EQ(finished, false);
+
+    auto num_fields = splits.size();
+    if (num_fields == 4u) {
+      //   0            1          2      3
+      // src_state  dest_state   label  score
+      int32_t src_state = StringToInt(splits[0]);
+      int32_t dest_state = StringToInt(splits[1]);
+      int32_t symbol = StringToInt(splits[2]);
+      float score = StringToFloat(splits[3]);
+      arcs.emplace_back(src_state, dest_state, symbol, score);
+    } else if (num_fields == 1u) {
+      //   0
+      // final_state
+      (void)StringToInt(splits[0]);  // this is a final state
+      finished = true;               // set finish
+    } else {
+      K2_LOG(FATAL) << "Invalid line: " << line
+                    << "\nK2 acceptor expects a line with 1 (final_state) or "
```

Comment:
BTW, we've been using lowercase k2 in text... this is open to discussion but I felt it would get complicated with directory names and variable names and the like if we consistently use uppercase.